pydantic/_internal/_typing_extra.py

```diff
@@ -265,19 +265,23 @@ def eval_type_backport(
                 value, globalns, localns
             )
     except TypeError as e:
-        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
-            raise
-        try:
-            from eval_type_backport import eval_type_backport
-        except ImportError:
-            raise TypeError(
-                f'You have a type annotation {value.__forward_arg__!r} '
-                f'which makes use of newer typing features than are supported in your version of Python. '
-                f'To handle this error, you should either remove the use of new syntax '
-                f'or install the `eval_type_backport` package.'
-            ) from e
-
-        return eval_type_backport(value, globalns, localns, try_default=False)
+        if isinstance(value, typing.ForwardRef):
```

Comment:
I feel like we haven't exactly preserved the logical flow here - don't we need a `is_backport_fixable_error(e))` check in the case where we jump straight to `raise e`?

Reply:
The check is present on line 269, however the `TypeError` message seems to have changed in Python 3.11:

Python 3.10: `TypeError: 'type' object is not subscriptable`

Python 3.11+: `TypeError: type 'CustomType' is not subscriptable`

and `is_backport_fixable_error` matches the first error message. This makes things even harder to manage unfortunately :/

Reply:
In other words, there's no way to distinguish between a `TypeError` when a builtin/stdlib class is used (e.g. `list[int]`, `collections.abc.Sequence[str]`) and other classes that are truly non subscriptable

Reply:
Ah gotcha. I'm ok with this approach then, but let's see what @alexmojaki thinks, as he's pretty well versed in the `eval_type_backport` space.

Reply:
I've pushed some changes which I think is the best we can do

Reply:
I suggest keeping `is_backport_fixable_error`, and also updating `is_not_subscriptable_error` to check for Python <= 3.8 and `is_unsupported_types_for_union_error` for Python <= 3.9 (double check those Python versions). The newer `type 'CustomType' is not subscriptable` format isn't backport fixable because the backport only applies to old versions. So you can definitely make a better guess about whether it's worth suggesting installing `eval_type_backport`.